### PR TITLE
[dev-tps-warping] Replace `_interpoint_distances` with SciPy's `cdist`

### PR DESCRIPTION
**Pull request by decorouz**
## Description
Replace `_interpoint_distances` function with built in `distance.cdist` function from Scipy spatial module.

@ana42742 @lagru  Please review this improvements??

## Checklist

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

### CODE REVIEW

**Comment by lagru** [comment]
Great suggestion @decorouz. `cdist` is even faster than `_interpoint_distances`. E.g. run this in ipython:

```python
from scipy.spatial.distance import cdist

rng = np.random.default_rng(42)
points = rng.random((500, 2))

def _interpoint_distances(points):
    xd = np.subtract.outer(points[:, 0], points[:, 0])
    yd = np.subtract.outer(points[:, 1], points[:, 1])
    return np.sqrt(xd**2 + yd**2)

res1 = _interpoint_distances(points)
res2 = cdist(points, points)
np.testing.assert_allclose(res1, res2)

%timeit _interpoint_distances(points)
%timeit cdist(points, points)
# 3.21 ms ± 17.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# 397 µs ± 899 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

Though, it also seems like it is a complete replacement for the former function.

```python
def _interpoint_distances(points):
    ds = (
        np.subtract.outer(points[:, i], points[:, i])
        for i in range(points.shape[1])
    )
    return np.sqrt(sum(d**2 for d in ds))
```
this even succeeds for more than two dimensions. Might be cool to look into whether this algorithm works for ND at some point.

I am tempted to merge this, but also think we should maybe be patient and wait until we have some basic test functions to verify that everything behaves as before...

**Comment by lagru** [comment]
Note that you seem to be making this PR from your fork's version of `dev-tps-warping` and not from a feature branch. You may now run into merge conflicts when you try to update that branch with changes from other PRs. E.g. https://github.com/scikit-image/scikit-image/commit/91862edf91fdb86bd76db425797ef650eaa23431 was just merged.

**Comment by decorouz** [comment]
I was able to update my branch with changes from this commit [91862ed](https://github.com/scikit-image/scikit-image/commit/91862edf91fdb86bd76db425797ef650eaa23431) with the command `git pull upstream dev-tps-warping`.

**Comment by lagru** [comment]
Huh, that's weird. Did you push the update to? For me it shows that your remote branch is 1 commit ahead (this PR) and 1 commit behind (https://github.com/scikit-image/scikit-image/commit/91862edf91fdb86bd76db425797ef650eaa23431). Git should have complained when running your command? :shrug: But I can live with that for now. :sweat_smile:

**Comment by decorouz** [comment]
> Huh, that's weird. Did you push the update to? For me it shows that your remote branch is 1 commit ahead (this PR) and 1 commit behind ([91862ed](https://github.com/scikit-image/scikit-image/commit/91862edf91fdb86bd76db425797ef650eaa23431)). Git should have complained when running your command? 🤷 But I can live with that for now. 😅

Yes, I pushed to update and git didn't complain yet? ;)
Me too, I can live with no git complains for now
 😅

File: skimage/transform/_thin_plate_splines.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import numpy as np
 from scipy import ndimage
+import scipy as sp
```
Comment (lagru) [inline skimage/transform/_thin_plate_splines.py:3]: I think @mkcor meant to replace the above line with the new import statement. You'll also need to update `ndimage.map_coordinates`, and `distance.cdist` below.

File: skimage/transform/_thin_plate_splines.py [hunk 2]
```diff
@@ -145,3 +137,4 @@ def _make_warp(from_points, to_points, x_vals, y_vals):
     y_warp = _calculate_f(coeffs[:, 1], from_points, x_vals, y_vals)
     np.seterr(**err)
     return [x_warp, y_warp]
+    return [x_warp, y_warp]
```
Comment (lagru) [inline skimage/transform/_thin_plate_splines.py:140]:
```suggestion

```

**Comment by decorouz** [comment]
@lagru @ana42742 Please review recent changes and docstring implementation.

**Comment by lagru** [comment]
Wow, @decorouz. I haven't had time yet to look this through, but the docstrings look very promising. :ok_hand:

File: skimage/transform/_thin_plate_splines.py [hunk 3]
```diff
@@ -97,51 +97,122 @@ def _make_inverse_warp(
...
+    ----------
+    x: ndarray
+        Input array representing the norm distance between points.
+        The norm is the Euclidean distance.
+    Returns
+    -------
+    ndarray
+        Calculated U values.
+    """
+    _small = 1e-8  # Small value to avoid divide-by-zero
```
Comment (ana42742) [inline skimage/transform/_thin_plate_splines.py:113]: Hmm, I am curious about why you decided to change the `_small` value.

Comment (decorouz) [reply skimage/transform/_thin_plate_splines.py:113]: You mean where it is declared or the value itself?
The value itself can be any small value. The variable isn't used in other functions, it appropriate the declare inside the function for closure and reduce noise in the global namespace.

File: skimage/transform/_thin_plate_splines.py [hunk 4]
```diff
@@ -66,82 +106,183 @@ def _make_inverse_warp(
...
+        # t10 = transform[0][(ix1, y_indices)]
+        # t11 = transform[0][(ix1, iy1)]
+        # transform_x = (t00*x1*y1 + t01*x1*y_fracs
+        #                + t10*x_fracs*y1 + t11*x_fracs*y_fracs)
+        # t00 = transform[1][(x_indices, y_indices)]
+        # t01 = transform[1][(x_indices, iy1)]
+        # t10 = transform[1][(ix1, y_indices)]
+        # t11 = transform[1][(ix1, iy1)]
+        # transform_y = (t00*x1*y1 + t01*x1*y_fracs
+        #                + t10*x_fracs*y1 + t11*x_fracs*y_fracs)
```
Comment (lagru) [inline skimage/transform/_thin_plate_splines.py:144]:
```suggestion
```

**Pull request merged**